### PR TITLE
[7.0.x] JBPM-5486: Stunner - persist magnet locations for connections

### PR DESCRIPTION
**Pull request by jlindop**
@romartin, @hasys Will you review for 7.0.x. It's on the 7.0.0.Final list https://issues.jboss.org/browse/JBPM-5486?filter=12331527

### CODE REVIEW

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/connection/ConnectionAcceptorControlImpl.java [hunk 1]
```diff
@@ -128,42 +136,64 @@ public boolean allowTarget(final Node target,
...
-            ensureUnHighLight();
             final CommandResult<CanvasViolation> violations = getCommandManager().execute(canvasHandler,
                                                                                           canvasCommandFactory.setSourceNode(source,
                                                                                                                              connector,
-                                                                                                                             magnet));
+                                                                                                                             magnet,
+                                                                                                                             magnetX,
+                                                                                                                             magnetY));
+            return isAccept(violations);
+        } else {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/connection/ConnectionAcceptorControlImpl.java]: code duplication on the `if/else` statements

File: kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/connection/ConnectionAcceptorControlImpl.java [hunk 2]
```diff
@@ -128,42 +136,64 @@ public boolean allowTarget(final Node target,
...
-            ensureUnHighLight();
             final CommandResult<CanvasViolation> violations = getCommandManager().execute(canvasHandler,
                                                                                           canvasCommandFactory.setTargetNode(target,
                                                                                                                              connector,
-                                                                                                                             magnet));
+                                                                                                                             magnet,
+                                                                                                                             magnetX,
+                                                                                                                             magnetY));
+            return isAccept(violations);
+        } else {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-lienzo/src/main/java/org/kie/workbench/common/stunner/client/lienzo/canvas/controls/connection/ConnectionAcceptorControlImpl.java]: code duplication on the `if/else` statements

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/SetConnectionSourceMagnetCommand.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+import org.kie.workbench.common.stunner.core.graph.content.view.View;
+import org.kie.workbench.common.stunner.core.graph.content.view.ViewConnector;
+import org.kie.workbench.common.stunner.core.rule.RuleViolation;
+import org.uberfire.commons.validation.PortablePreconditions;
+
+/**
+ * A Command to set the magnet for a node
+ */
+@Portable
+public final class SetConnectionSourceMagnetCommand extends AbstractGraphCommand {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/SetConnectionSourceMagnetCommand.java]: Code duplication as in `SetConnectionSourceNodeCommand`

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/SetConnectionTargetMagnetCommand.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+import org.kie.workbench.common.stunner.core.graph.content.view.View;
+import org.kie.workbench.common.stunner.core.graph.content.view.ViewConnector;
+import org.kie.workbench.common.stunner.core.rule.RuleViolation;
+import org.uberfire.commons.validation.PortablePreconditions;
+
+/**
+ * A Command to set the magnet for a node
+ */
+@Portable
+public final class SetConnectionTargetMagnetCommand extends AbstractGraphCommand {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/SetConnectionTargetMagnetCommand.java]: Code duplication as in `SetConnectionTargetNodeCommand`

File: kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/marshall/json/builder/AbstractNodeBuilder.java [hunk 5]
```diff
@@ -242,14 +251,65 @@ private boolean doExecuteCommand(final BuilderContext context,
...
-        return 7;
+                                              final Edge<ViewConnector<W>, Node> edge,
+                                              Double[] docker) {
+        int magnetIndex = calculateMagnetIndex(node,
+                                               docker,
+                                               7);
+        return magnetIndex;
+    }
+
+    private int calculateMagnetIndex(final T node,
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/marshall/json/builder/AbstractNodeBuilder.java]: would avoid calculating magnet indexes on server side. This code depends on the the shape state on client side and should not be part of the domain model.

**Comment by hasys** [comment]
I see the difference between current Designer and Stunner when you attach arrow to center of node:
Stunner:
image
Current Designer:
image

IMHO behavior of current Designe is better.

Also am I correct, that this PR fix "Missing persistence for the different connections magnets" and don't fix "Missing persistence for the different connectors control points as well" for now?

**Comment by jlindop** [comment]
@hasys I like the way designer handles magnets - on serialization, it saves the mid-point of the node as the point that the connector goes to, then on deserialisation it attaches the connector to the nearest side. (I also like the diagonal lines in Designer). In Stunner, if users attach connectors to the middle magnet, then we'll serialize it and deserialize it, as the user must have deliberately attached them there.

You're right in thinking that this PR only addresses the source and target magnets and not control points. I've created a linked Jira for the control points. 

I'm currently re-working the code according to Roger's suggestions and I'll incorporate yours too. Thanks.

**Comment by hasys** [comment]
@jlindop ok, thank you! 

So, we need to create addition Jira issue about magnets handlers. Did I understand correct?

**Comment by jlindop** [comment]
@hasys I've created the linked Jira https://issues.jboss.org/browse/JBPM-5962 for control points.

**Comment by romartin** [comment]
thanks @jlindop @hasys , good point both! Yeah let's do it by steps, let's let @jlindop close this serialization for magnets, then we can address the [JBPM-5962](https://issues.jboss.org/browse/JBPM-5962) as next, to improve this behavior. Sounds good for me, thanks!

**Comment by jlindop** [comment]
@romartin @hasys I've pushed the changes to address your comments. Will you take another look at this commit? Thanks.

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/graph/content/view/Magnet.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.workbench.common.stunner.core.graph.content.view;
+
+public interface Magnet {
+
+    enum MagnetType {
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/graph/content/view/Magnet.java]: hey @jlindop , not sure why need the `MagnetType`..  so having a connector instance, it has two magnets associated on the model side - the source and target ones. So the information that `MagnetType` provides it's  already provided just by the connector's content (`ViewConnector`), depending on if you're obtaining the magnet from the source or the target nodes.
I see that you're using this for the `WiresViewConnector` class, but i would say that it's not really necessary - the Magnet instance comes itself from a source or target shape, which comes at same time from a edge instance, so this info is implicit... isn't it?

File: kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/SetConnectionSourceNodeCommand.java [hunk 7]
```diff
@@ -44,33 +45,40 @@
 
     private final String sourceNodeUUID;
     private final String edgeUUID;
-    private final Integer magnetIndex;
+    private final Magnet magnet;
+    // true if a new connection is being made; false if a different magnet is being selected
+    // on an existing connection
+    private final boolean isNewConnection;
```
Comment (romartin) [inline kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-core-common/src/main/java/org/kie/workbench/common/stunner/core/graph/command/impl/SetConnectionSourceNodeCommand.java]: hey @jlindop , why you need this `isNewConnection` flag here? In the 'ViewConnector' class, the source and target magnets are returning `Optional` instances, is it not enough in order to figure out this flag value?

**Pull request merged**